fix failing concatenate and add guard condition

hokiegeek2 opened the pull request:
This PR does the following:
1. Addresses #721 by disabling the testConcatenate test case that fails due to #710 
2. Adds client-side guard condition that prevents manifestation of #710 by checking to see if any Categorical or Strings object to be concatenated has a len() < numLocales. If so, a ValueError is raised.
3. Add test_concatenate to string_test

reuster986 commented:
@hokiegeek2 I don't see the guard condition you added -- the only file I see changed is `string_test.py`. Am I missing something?

More to the point, I don't think the guard condition should raise a `ValueError`, given that concatenating small arrays is a perfectly valid operation. Better to have the guard condition call `.to_ndarray()`, do the concatenation on the client, then transfer the result back over to the server. 

Finally, I just want to make sure we are treating this PR as a temporary workaround while we continue working to get concatenation correct across all cases.

hokiegeek2 commented:
> @hokiegeek2 I don't see the guard condition you added -- the only file I see changed is `string_test.py`. Am I missing something?
> 
> More to the point, I don't think the guard condition should raise a `ValueError`, given that concatenating small arrays is a perfectly valid operation. Better to have the guard condition call `.to_ndarray()`, do the concatenation on the client, then transfer the result back over to the server.
> 
> Finally, I just want to make sure we are treating this PR as a temporary workaround while we continue working to get concatenation correct across all cases.

Argh, sorry, git commit fail on my part, fixed. 

The guard condition is there because concatenation of Strings or Categorical objects where 1..n of the objects have a length < numLocales fails. I thought I had fixed this in #710 but the tests you added proved that my change caused other errors. The guard condition does not apply to pdarrays.

And yes, as I discussed with @mhmerrill, this is intended to be a temp workaround.

reuster986 reviewed (approved):
Looks like users can still concatenate small arrays using `ordered=True`, so I'm fine with this for now, although maybe the error message should tell the user to retry with `ordered=True`.

hokiegeek2 commented:
> Looks like users can still concatenate small arrays using `ordered=True`, so I'm fine with this for now, although maybe the error message should tell the user to retry with `ordered=True`.

yes, I apologize, I should have made that clearer. The guard condition only applles to Strings or Categoricals in interleave mode AKA ordered=False. I will update the error message to reflect that

reuster986 commented:
> The guard condition is there because concatenation of Strings or Categorical objects where 1..n of the objects have a length < numLocales fails. I thought I had fixed this in #710 but the tests you added proved that my change caused other errors. The guard condition does not apply to pdarrays.

I was talking about Categoricals, not pdarrays. Side note: when I say "array", I mean any of the following: pdarray, Strings, Categorical, Timedelta, Datetime -- basically anything that can be a column of data. I wanted to make sure the user has a way to concatenate small Categorical arrays, but it looks like they can still do that with `ordered=True` without needing to bring back to the client -- is that correct?

hokiegeek2 commented:
> > The guard condition is there because concatenation of Strings or Categorical objects where 1..n of the objects have a length < numLocales fails. I thought I had fixed this in #710 but the tests you added proved that my change caused other errors. The guard condition does not apply to pdarrays.
> 
> I was talking about Categoricals, not pdarrays. Side note: when I say "array", I mean any of the following: pdarray, Strings, Categorical, Timedelta, Datetime -- basically anything that can be a column of data. I wanted to make sure the user has a way to concatenate small Categorical arrays, but it looks like they can still do that with `ordered=True` without needing to bring back to the client -- is that correct?

👍 Indeed!

reuster986 commented:
@hokiegeek2 looks like you just answered my question -- thanks!

hokiegeek2 commented:
@reuster986 new error message:

![image](https://user-images.githubusercontent.com/10785153/115391536-92d85f80-a1ad-11eb-8f2f-8f67f0a46cee.png)

reuster986 commented:
@mhmerrill would you like to review, or should we go ahead and merge?
